File: 0189-rotate-array/solution.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n; // reduce k if larger than n
        int count = 0; // number of elements moved

        for (int start = 0; count < n; start++) {
            int current = start;
            int prev = nums[start];

            do {
                int nextIdx = (current + k) % n;
                int temp = nums[nextIdx];
                nums[nextIdx] = prev;
                prev = temp;
                current = nextIdx;
                count++;
            } while (start != current); // stop when cycle completes
        }
    }
}

File: 0189-rotate-array/solution.py
class Solution(object):
    def rotate(self, nums, k):
        """
        n = len(nums)
        k %= n
        for _ in range(k):
            last = nums[-1]
            for i in range(n):
                nums[i], last = last, nums[i]
        """
        n = len(nums)
        k %= n
        nums.reverse()
        nums[:k] = reversed(nums[:k])
        nums[k:] = reversed(nums[k:])

